### ISSUE TITLE
Decide on the structure of the `Device` classes

### ISSUE THREAD

**Issue by teaguetomesh** [teaguetomesh/c2qa-modular-qc#7]
My original thought was the represent a `Device` using a `NetworkX.Graph` where the nodes of that graph represent individual modules. However, for compiling circuits it is useful to have a [CouplingMap](https://qiskit.org/documentation/stubs/qiskit.transpiler.CouplingMap.html) defined which labels each qubit of the device individually.

It is up to us how we want to go about labelling the modules and qubits of the distributed devices, but I think it'd be better to think about this design choice early on rather than waiting until later and requiring a bunch more work to change anything.

I see a few possible choices:

1. The `Device`s  represent the module coupling using a `NetworkX.Graph` where the indices of the graph represent the different modules: `node_1 -> module_1`, etc. We still have a unique label for every qubit in the device, it will just require two indices. For example, the 5th qubit of the 3rd module would be `my_device.modules[2].qubits[4]`. Then we would just need to write a function that takes in a `Device` and spits out a `CouplingMap` (if we want to use Qiskit's compiler functions).

One issue I see with this approach is that the "device graph" only contains information about which modules are connected to each other  (for example, "module 1 is connected to module 2") -- it doesn't have finer grained details such as: qubit 1 of module 1 is connected to qubit 0 of module 2. The finer grained detail is what is needed for the `CouplingMap`.

2. The `Device` class explicitly stores the whole coupling map as one of its class variables. In this case I don't really see what the use of the `Module` classes would be beyond acting as constructors when the `Device` is building the coupling map.

3. The `Device` class still constructs a `NetworkX.Graph` where the nodes represent modules, but we change how the modules are constructed so that we can build a module that contains qubits 0-5, and the next module contains qubit 6-10. Basically, instead of constructing the module using `num_qubits` we would just pass in a list of `qubits`. This would help us to keep track of the physical qubit labelling and make it easy to construct the `CouplingMap`. The one thing we would still need is a way to represent edges between qubits of different modules. I think this can be done by creating a dictionary `Device.module_interconnects` or something like that where the keys of the dictionary would be the edges of the device graph and the values would be the qubit indices.

question: Are there any other things we need out of a `Device` aside from a `CouplingMap` that I am missing?

**Comment by weiT1993** [comment]
The `Device` class will just take two graphs - a device graph for the inter-module connections and a list of graphs for each individual modules. As a starting point, the list of module graphs will be the same as we assume homogeneous modules. But this leaves room for heterogenous if needed.

The `Device` will also have two sets of edges - global and local (complete? full?). Global are just the edges in between the modules. Local will be all the edges both inter and intra. Local edges information is more for benchmarking Qiskit compilers rather than developing our distributed compiler.

**Comment by teaguetomesh** [comment]
Yeah that makes sense. I think we will still need to make a few changes to how the `Device` and `Module` classes are structured right now. I will work on implementing the change I suggested in the first sentence of number 3 above - it'll be useful if we can construct individual modules with uniquely named qubits.

And then the list of Global edges will still need to be more specific than just "module 1 is connected to module 2". It will also need finer grained info like "qubit 1 of module 1 is connected to qubit 6 of module 2".

**Comment by weiT1993** [comment]
Define qubits in modular devices as a whole. E.g. module 0 has qubits 0-9, module1 has qubits 10-19 etc.

There are three sets of edges definitions for various purposes:

1. Global edges in between the modules, represented by the module indices. Used to generate the target architecture file for SCOTCH.
2. Local edges within each module represented by physical qubits. Used for local compile.
3. All edges represented by physical qubits. Used for global communications and profile Qiskit.

**Comment by weiT1993** [comment]
One caveat is that Qiskit `transpile` only takes `coupling_map` as integers starting from 0. 

```
circuit = generate_circ(num_qubits=6,depth=1,circuit_type='regular',reg_name='haha',connected_only=False,seed=None)
edges = [[0,1],[1,2],[2,3],[3,4],[4,5]]
offset = 0
coupling_map = CouplingMap()
[coupling_map.add_physical_qubit(qubit) for qubit in range(offset+6)]
for edge in edges:
    coupling_map.add_edge(edge[0]+offset,edge[1]+offset)
    coupling_map.add_edge(edge[1]+offset,edge[0]+offset)
print(coupling_map)
transpiled_circuit = transpile(circuit,coupling_map=coupling_map,layout_method='sabre',routing_method='sabre')
print(transpiled_circuit)
```

produces `[[0, 1], [1, 0], [1, 2], [2, 1], [2, 3], [3, 2], [3, 4], [4, 3], [4, 5], [5, 4]]` and a successful transpilation. Setting `offset` to nonzero complains that the coupling map is not connected.

The implication for us is that we have to label module qubits for each module starting from 0, which I already did. But this also means we most likely have to have a physical qubit to physical qubit translation.

So the device qubits will be labeled 0->49 (for example), and module 0 is 0_0->0_9, module 1 is 1_0->1_9 etc. And we need a mapping from 0->49 to the 1_0 etc module qubits.

**Comment by weiT1993** [comment]
Qubit definition example for a device with 5 modules, each with 10 qubits. Assume the input circuit is also 50 qubits.
- `device_physical`: labeled 0-49 int.
- `device_virtual`: `Qubit(QuantumRegister(50, 'q'), i)`, where `i=0-49`.
- `module_physical`: labeled 0-9 int for each module.
- `module_virtual`: `Qubit(QuantumRegister(10, 'q'), i)`, where `i=0-9`.

The translations needed:
- `dp_2_dv_mapping`: Dict. Keys: `device_physical`. Values: `device_virtual`.
- `mp_2_mv_mapping`: Dict. Keys: `module_physical `. Values: `module_virtual `.
- `dp_2_mp_mapping`: Dict. Keys: `device_physical `. Values: `module_physical `.
Each translation also has their reverse dict.

To benchmark existing monolithic compiler, use `device_physical` as the coupling map and `device_virtual` as the circuit qubits.

To develop modular compiler, use `module_physical` as the coupling map and `module_virtual` as the local circuit qubits for local compile. Further use the translations to combine the local circuits into the global circuit.

**Comment by teaguetomesh** [comment]
> One caveat is that Qiskit transpile only takes coupling_map as integers starting from 0.

Ah, okay yeah it says it right in the definition as well: https://qiskit.org/documentation/stubs/qiskit.transpiler.CouplingMap.html
That's annoying but whatever, that decides our implementation for us.
